### PR TITLE
#749 Bad error handling for IOException while reading incoming request body

### LINKED ISSUE #749: Bad error handling for IOException while reading incoming request body
## Expected Behavior

If an `IOException` happens while reading the incoming request body (in `RequestMapper.MapContent` [here](https://github.com/ThreeMammals/Ocelot/blob/02e5cea7b1bdd3d78cfe1a5bb688383f54a809fc/src/Ocelot/Request/Mapper/RequestMapper.cs#L24)), Ocelot fails the request with an `UnmappableRequestError` and status code 404.

404 is not appropriate, and the error logged by ocelot should be more specific than `UnmappableRequestError`. It just so happens that the exception happens while doing request mapping, but it's really an IO problem while reading the incoming request stream. Response status shouldn't be 404. In the case of too large request, 400 might be appropriate. In other cases, something in the 5xx range.

## Steps to Reproduce the Problem

This happens consistently, for example, for requests that are too large for the server (`HttpSys` throws an IOException if the Content-Length of the incoming request is larger than 30MB by default -- see: [AspNetCore source](https://github.com/aspnet/AspNetCore/blob/cd0eab88eaa230fa276c27ab5dc71ea267efe14f/src/Servers/HttpSys/src/RequestProcessing/RequestStream.cs#L435-L436)).

  1. Send a request to Ocelot, hosted on HttpSys, with Content-Length larger than 30,000,000.
  1. Boom!

Ocelot returns 404, and logs `UnmappableRequestError`.

## Specifications

  - Version: 12.0.1 (should also apply to 13.0.0 by code inspection)
  - Platform: Windows Server 2016 / .NET Framework 4.6.2

### PR DESCRIPTION
**Pull request by ks1990cn**
## Fixes #749 
- #749 
- Fixing issue for status code 404, when content length is large than body size. will be displaying **413** status code ("Entity Payload too large"). 
- Considered scenario when application is hosted on Kestrel , IIS or Http.Sys and have different options configured as per host server.

**For example**:
```c#
builder.WebHost.UseHttpSys(options =>
{
    options.MaxRequestBodySize = 60000000;
});
```
Above code have `MaxRequestBodySize  `configured manually (default is ~28.6MB), so changes will handle this case too.

## Proposed Changes
- I have done manual testing, writing test case to fake server host was tough, as per my research on ASP.NET Core there is no exposed way to directly write this fake server test.
- I also faced issue with access level of `ToByteArray` of `RequestMapper`, so I could have marked it null and `ByteArrayContent` constructor have given me error and hence I could have tested it.
- Open for discussion how we can test this functionality :D

### CODE REVIEW

**Comment by raman-m** [comment]
Tanmay,
The issue #729 is about sticky sessions! Are you sure your PR is related to #729 ?
It seems the number is wrong...

**Comment by raman-m** [comment]
I found it! The issue is #749 😺 
- #749

**Comment by ks1990cn** [comment]
> I found it! The issue is #749 😺
> 
> * [Bad error handling for IOException while reading incoming request body #749](https://github.com/ThreeMammals/Ocelot/issues/749)

Sorry! correct its 749. Thank you

**Review by ggnaegi — CHANGES_REQUESTED** [review]
Hello

I think, you could improve the code a bit. I would recommend creating a service for the exception handling, otherwise you are mixing concerns in the RequestMapper class.

You will be then able to write some unit tests for your service.

Acceptance tests are missing.

File: src/Ocelot/Request/Mapper/RequestMapper.cs [hunk 1]
```diff
@@ -1,59 +1,122 @@
...
+                if (PayloadTooLargeOnAnyHostedServer(request, ex))
+                {
+                    return new ErrorResponse<HttpRequestMessage>(new PayloadTooLargeError(ex));
+                }
+
+                return new ErrorResponse<HttpRequestMessage>(new UnmappableRequestError(ex));
+            }
+        }
+
+        private bool PayloadTooLargeOnAnyHostedServer(HttpRequest request, Exception ex)
```
[the remark this answers is not in the document]
Comment (ks1990cn) [reply src/Ocelot/Request/Mapper/RequestMapper.cs]: Acknowledged and made related changes.

File: src/Ocelot/Request/Mapper/RequestMapper.cs [hunk 2]
```diff
@@ -1,59 +1,122 @@
...
-        private static async Task<HttpContent> MapContent(HttpRequest request)
+        private readonly long? maxRequestBodySizeHttpsSys;
+        private readonly long? maxRequestBodySizeKerstrelServer;
+        private readonly long? maxRequestBodySizeIISServer;
+        private IServer _server;
+        public RequestMapper()
+        {
+        }
+
+        public RequestMapper(IOptions<HttpSysOptions> httpSysOptions, IOptions<KestrelServerOptions> kerstrelServerOptions, IOptions<IISServerOptions> iISOptions,IServer server)
```
[the remark this answers is not in the document]
Comment (ks1990cn) [reply src/Ocelot/Request/Mapper/RequestMapper.cs]: Acknowledged and made related changes.

**Comment by ggnaegi** [comment]
@ks1990cn @raman-m  Hello, I took now a bit of time to create an acceptance test, and I realized that the implementation is WAAAAAAAAAAAAAY too complicated (and it's impossible to get the KestrelServer from IServer, because you will get KestrelServerImpl first). Kestrel and other servers will return you a ```BadHttpRequestException``` with status code 413. So the issue here, is that the exception is wrapped and hidden in the UnmappableRequestError.  A rethinking of the exception handling is needed... 

The issue here can be easily be fixed with:
```c#
catch (BadHttpRequestException badHttpRequest)
            {
                return badHttpRequest.StatusCode == 413 ? new ErrorResponse<HttpRequestMessage>(new PayloadTooLargeError(badHttpRequest)) : new ErrorResponse<HttpRequestMessage>(new UnmappableRequestError(badHttpRequest));
            }
            catch (Exception ex)
            {
                return new ErrorResponse<HttpRequestMessage>(new UnmappableRequestError(ex));
            }
```
Again, it's only a patch, imho, the exception should bubble up to a central exception handler.
The acceptance test:
``` c#
[Fact]
        public void should_return_status_413_when_exceeding_body_max_size()
        {
            var port = RandomPortFinder.GetRandomPort();

            var configuration = new FileConfiguration
            {
                Routes = new List<FileRoute>
                {
                    new()
                    {
                        DownstreamPathTemplate = "/",
                        UpstreamPathTemplate = "/",
                        UpstreamHttpMethod = new List<string> { "Post" },
                        DownstreamHostAndPorts = new List<FileHostAndPort>
                        {
                            new()
                            {
                                Host = "localhost",
                                Port = port,
                            },
                        },
                        DownstreamScheme = "http",
                    },
                },
            };

            this.Given(x => x.GivenThereIsAServiceRunningOnNoExceptions($"http://localhost:{port}"))
                .And(x => _steps.GivenThereIsAConfiguration(configuration))
                .And(x => _steps.GivenOcelotIsRunningOnKestrelWithCustomBodyMaxSize(1024))
                .When(x => _steps.WhenIPostUrlOnTheApiGateway("/", new ByteArrayContent(File.ReadAllBytes(@"C:\Users\gnaegu\Desktop\test45.txt"))))
                .Then(x => _steps.ThenTheStatusCodeShouldBe(413))
                .BDDfy();
        }
```
Kestrel Server Implementation
```c#
public void GivenOcelotIsRunningOnKestrelWithCustomBodyMaxSize(long customBodyMaxSize)
        {
             var kestrelServer = Host.CreateDefaultBuilder()
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseKestrel().ConfigureKestrel((_, options) =>
                        {
                            options.Limits.MaxRequestBodySize = customBodyMaxSize;
                        })
                        .ConfigureAppConfiguration((hostingContext, config) =>
                        {
                            config.SetBasePath(hostingContext.HostingEnvironment.ContentRootPath);
                            var env = hostingContext.HostingEnvironment;
                            config.AddJsonFile("appsettings.json", optional: true, reloadOnChange: false)
                                .AddJsonFile($"appsettings.{env.EnvironmentName}.json", optional: true, reloadOnChange: false);
                            config.AddJsonFile(_ocelotConfigFileName, optional: true, reloadOnChange: false);
                            config.AddEnvironmentVariables();
                        })
                        .ConfigureServices(s =>
                        {
                            s.AddOcelot();
                        })
                        .Configure(app =>
                        {
                            app.Use(async (_, next) =>
                            {
                                await next.Invoke();
                            });
                            app.UseOcelot().Wait();
                        })
                        .UseUrls("http://localhost:5001");
                }).Build();
             kestrelServer.Start();

             _ocelotClient = new HttpClient
             {
                 BaseAddress = new Uri("http://localhost:5001"),
             };
        }
```

**Review by ggnaegi — CHANGES_REQUESTED** [review]
Please look at my comment. The implementation is overkill, this can be easily be fixed.

**Comment by raman-m** [comment]
Guys, who is our hero?
I'll get back to this funny thread after current release...
Keep conversation ongoing! ✌️😁

**Comment by ggnaegi** [comment]
> Guys, who is our hero?
> I'll get back to this funny thread after current release...
> Keep conversation ongoing! ✌️😁

Well, not really. I'm just tired of this $%^%&*% error handling.

**Comment by ks1990cn** [comment]
Ah! didn't noticed that it already gives 413 in exception, I remember it was giving me 404 only (Maybe I didn't noticed right). Correct it can be done directly.
If you are ready @ggnaegi, you can go ahead. Converting it to draft. Thanks for taking look @ggnaegi .

**Comment by raman-m** [comment]
Open.. close... open... close... I'm going crazy with you guys 😃 
What is the follow up after closing this PR?
What is the result?

**Pull request closed without merging**